proposals/0000-unstable.rst

```diff
@@ -0,0 +1,306 @@
+     GHC.Debug.debugLn
+     GHC.Debug.debugErrLn
+     GHC.Magic.oneShot
+     GHC.Prim.Ext.getThreadAllocationCounter#
+     GHC.Prim.Ext.asyncRead#   -- only on mingw32
+     GHC.Prim.Ext.asyncWrite#  -- only on mingw32
+     GHC.Prim.Ext.asyncDoProc# -- only on mingw32
+     GHC.Prim.Panic.absentSumFieldError
+     GHC.Prim.Panic.panicError
+     GHC.Prim.Panic.absentError
+     GHC.Prim.Exception.raiseOverflow   -- other similar functions are exported by "ghc-bignum".GHC.Num.Primitives
+     GHC.Tuple.getSolo                  -- other similar definitions are exported by "base".GHC.Tuple
+     GHC.Types.Module                   -- other similar definitions are exported by "base".Type.Reflection.Unsafe
+     GHC.Types.TrName                   -- other similar definitions are exported by "base".Type.Reflection.Unsafe
+     GHC.Types.TyCon                    -- other similar definitions are exported by "base".Type.Reflection.Unsafe
+     GHC.Types.KindBndr                 -- other similar definitions are exported by "base".Type.Reflection.Unsafe
+
+   Any user of any of these exports will have no backward-compatible way forward if
+   this proposal is accepted and implemented, other than specifying ``-XUnstable``
+   (which isn't backward compatible). If we believe that any of the above functions
+   are used beyond GHC itself, we should introduce a migration period.
+
+#. The ``FUN`` type is currently exported from ``GHC.Exts``, but this export would
+   be removed as part of this proposal. Accordingly, any user of ``FUN`` will have
+   a hard breakage, where they have to now import ``GHC.Prim`` (to get ``FUN``) and
+   have to use CPP to specify ``-XUnstable``.
+
+   ``FUN :: forall (m :: Multiplicity) -> forall {r1 :: RuntimeRep} {r2 :: RuntimeRep}. TYPE r1 -> TYPE r2 -> Type``
+   is the primitive function type, but its current kind is likely to change.
+
+Costs and Drawbacks
+-------------------
+
+1. Some have argued that GHC language extensions should be a way of guarding features
+   that have not yet made it into the next language standard. Instead, the ``-XUnstable``
+   and ``-XStable`` extensions are more like configuration options than language extensions.
+   Accepting this proposal moves us further from the vision that some future version of
+   Haskell will subsume all extensions.
+
+#. The naming of these extensions may be confusing. In particular, users may wonder
+   what the difference between ``-XNoUnstable`` and ``-XStable`` is, or think that
+   (the non-existent) ``-XNoStable`` is the same as ``-XUnstable``.
+
+Alternatives
+------------
+
+1. Instead of making a new language extension, we could imagine a new warning ``-Wunstable-features``.
+   This warning would be werror-by-default (no warnings are like this today) but could be disabled
+   with ``-Wno-unstable-features``. Similarly, ``-XStable`` could become something like ``-fstable``,
+   but this seems harder to specify in a ``.cabal`` file.
```

Comment:
This is an interesting point. The `nix` tool guard additional subcommand (a bit like extensions) under an “enable experimental features” configuration option.

So using a flag `-fenable-experimental-features` or `-Wno-unstable-features` is an alternative worth discussing.

I don’t think it’s harder to specify this in the `.cabal` file`, `ghc-options: -Wno-unstable-features`. People already know how to set warning flags in the cabal file. And those who don’t know are certainly not the target audience for this flag.